src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -193,51 +218,47 @@ internal ItemManager(RecentItemsManager recentItemsManager)
                         initialRoslynTriggerKind == CompletionTriggerKind.Invoke ||
                         filterText.Length <= 1)
                     {
-                        initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false)));
+                        builder.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false, patternMatch)));
                     }
                 }
             }
 
-            // DismissIfLastCharacterDeleted should be applied only when started with Insertion, and then Deleted all characters typed.
-            // This confirms with the original VS 2010 behavior.
-            if (initialRoslynTriggerKind == CompletionTriggerKind.Insertion &&
-                data.Trigger.Reason == CompletionTriggerReason.Backspace &&
-                completionRules.DismissIfLastCharacterDeleted &&
-                session.ApplicableToSpan.GetText(data.Snapshot).Length == 0)
-            {
-                // Dismiss the session
-                return null;
-            }
-
-            if (initialListOfItemsToBeIncluded.Count == 0)
+            if (builder.Count == 0)
             {
                 return HandleAllItemsFilteredOut(reason, data.SelectedFilters, completionRules);
             }
 
-            var options = document?.Project.Solution.Options;
-            var highlightMatchingPortions = options?.GetOption(CompletionOptions.HighlightMatchingPortionsOfCompletionListItems, document.Project.Language) ?? true;
+            var experimentationService = document.Project.Solution.Workspace.Services.GetService<IExperimentationService>();
+
+            var initialListOfItemsToBeIncluded = experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SortCompletionListByMatch)
+                // Need a stable sorting algorithm to preserve the original order for items with same pattern match score.
+                ? builder.OrderBy(new NullablePatternMatchComparer()).ToImmutableArray()
```

Comment:
Nullable?

Reply:
1. still need to renme this type.
2. can you not put comments *within* an expression?

Reply:
you are producin a new list by calling OrderBy, which defeats the purpose of the builder.  You can just call `.Sort` on hte builder.  Then you can use ToImmutableAndFree for both.

Reply:
I used `Builder.Sort` first but noticed it isn't stable (thus the comment). Let me confirm this.

Reply:
Confirmed `Builder.Sort` isn't stable.

Reply:
1. jfc
2. is tehre additional info we can add to make the items stable?  i.e. some sort of monotonically increasing integer they compare if all other things are equal? :)

Reply:
note the stability only matter in as far as normal vs import completion right? so if they were equal, we'd just see if this was an import completion item and order that second, right?

Reply:
> jfc
I sure was surprised.

> is tehre additional info we can add to make the items stable? 

Sure, I mean, I can add an integer to `ExtendedFilterResult` and make it monotonically increasing while creating them.

> note the stability only matter in as far as normal vs import completion right

The cause is sorting text and filter text being different, so we can't adjust the comparer just based on "normal vs import completion"

Reply:
Fixed